src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexNode.cs

```diff
@@ -97,25 +97,55 @@ public static RegexNode CreateOneWithCaseConversion(char ch, RegexOptions option
             {
                 Debug.Assert(culture is not null);
 
-                // If the character is part of a Unicode category that doesn't participate in case conversion,
-                // we can simply strip out the IgnoreCase option and make the node case-sensitive.
-                if (!RegexCharClass.ParticipatesInCaseConversion(ch))
+                // The only characters which may have different case-mappings depending on culture are 'i' | 'I' | '\u0130' | '\u0131'
+                // If the character being converted is one in that set, then we need to check current culture in order to use the right
+                // mappings. This special casing logic is duplicated in RegexCaseEquivalences.TryFindCaseEquivalencesForCharWithIBehavior
+                // with the reason being that we want to optimize for this case and efficiently return the string class which includes
+                // equivalences for the special case mappings.
+                if ((ch | 0x20) == 'i' || (ch | 0x01) == '\u0131')
                 {
-                    return new RegexNode(RegexNodeKind.One, options & ~RegexOptions.IgnoreCase, ch);
-                }
+                    RegexCaseBehavior mappingBehavior = RegexCaseEquivalences.GetRegexBehavior((options & RegexOptions.CultureInvariant) != 0 ? CultureInfo.InvariantCulture : culture);
+                    string? stringSet = ch switch
+                    {
+                        // Invariant mappings
+                        'i' or 'I' when mappingBehavior is RegexCaseBehavior.Invariant => "\0\u0004\0IJij",
 
-                // Create a set for the character, trying to include all case-insensitive equivalent characters.
-                // If it's successful in doing so, resultIsCaseInsensitive will be false and we can strip
-                // out RegexOptions.IgnoreCase as part of creating the set.
-                string stringSet = RegexCharClass.OneToStringClass(ch, culture, out bool resultIsCaseInsensitive);
-                if (!resultIsCaseInsensitive)
-                {
+                        // Non-Turkish mappings
+                        'i' or 'I' or '\u0130' when mappingBehavior is RegexCaseBehavior.NonTurkish => "\0\u0006\0IJij\u0130\u0131",
+
+                        // Turkish mappings
+                        'I' or '\u0131' when mappingBehavior is RegexCaseBehavior.Turkish => "\0\u0004\0IJ\u0131\u0132",
+                        'i' or '\u0130' when mappingBehavior is RegexCaseBehavior.Turkish => "\0\u0004\0ij\u0130\u0131",
+
+                        // Default
+                        _ => null
+                    };
```

Comment:
I think it'd be better to avoid this special logic here.  I realize you likely added it to reduce allocation for i's, which is a good goal, but this is non-trivial behavior that it'd be better to keep centralized in the one place that creates the equivalence sets.  I suggest just deleting this whole if block.